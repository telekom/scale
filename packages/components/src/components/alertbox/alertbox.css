:host {
  /* Token not available, up for discussion - Alerbox Variants/Colors*/
  --variant-error: hsla(353.79999999999995, 100%, 88.6%, 0.35);
  --variant-warning: hsla(33.5, 100%, 75.1%, 0.3);
  --variant-informational: hsla(190.20000000000005, 100%, 92%, 0.75);
  --variant-success: hsl(98.80000000000001, 49.7%, 61%, 0.2);

  /* Alertbox styling */
  --alertbox-border-radius: 8px;
  --alertbox-border: 1px solid var(--scl-color-white);
  --alertbox-width: 100%;
  --alertbox-font-header: var(--scl-fonts-family-sans);
  --alertbox-font-size-header: 16px;
  --alertbox-font-text: var(--scl-font-weight-regular);
  --alertbox-font-size-text: 16px;
}

.alertbox {
  border: var(--alertbox-border);
  border-radius: var(--alertbox-border-radius);
  width: var(--alertbox-width);
  position: relative;
}

/*Alertbox Color*/
.alertbox--variant-error {
  background: var(--variant-error);
  color: var(--scl-color-black);
}
.alertbox.alertbox--variant-warning {
  background-color: var(--variant-warning);
  color: var(--scl-color-black);
}
.alertbox.alertbox--variant-informational {
  background-color: var(--variant-informational);
  color: var(--scl-color-black);
}
.alertbox.alertbox--variant-success {
  background-color: var(--variant-success);
  color: var(--scl-color-black);
}

/*header*/
.alertbox .alertbox__container-header {
  font-weight: bold;
  margin: 14px 48px 14px 48px;
  font-family: var(--alertbox-font-header);
  font-size: var(--alertbox-font-size-header);
  line-height: 1.25;
}

/* Text/Content */
.alertbox .alertbox__content {
  margin: 8px 48px 15px 48px;
  font-family: var(--alertbox-font-text);
  font-size: var(--alertbox-font-size-text);
}

/* Close Icon/Button */
.alertbox .alertbox__container-header .alertbox__icon-close {
  position: absolute;
  top: 13px;
  right: 17px;
  color: black;
}

.alertbox .alertbox__container-header .alertbox__icon-close svg {
  height: 19px;
  width: 19px;
  padding: 3px;
  border-radius: 30%;
}

.alertbox .alertbox__container-header .alertbox__icon-close :hover {
  background-color: white;
  color: var(--scl-color-primary-active);
}

/* Success Icon */
.alertbox .alertbox__icon-success {
  position: absolute;
  top: 13px;
  left: 17px;
  height: 20px;
  width: 20px;

  color: var(--scl-color-background-success);
}

/* Error Icon */
.alertbox .alertbox__icon-error {
  position: absolute;
  top: 13px;
  left: 17px;
  height: 20px;
  width: 20px;
  color: var(--scl-color-background-error);
}

/* Info Icon */
.alertbox .alertbox__icon-information {
  position: absolute;
  top: 13px;
  left: 17px;
  height: 20px;
  width: 20px;
  color: var(--scl-color-text-info);
}

@media screen and (forced-colors: active), (-ms-high-contrast: active) {
  .alertbox__icon-close {
    color: white;
  }
}
