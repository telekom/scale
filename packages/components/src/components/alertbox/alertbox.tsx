import {
  Component,
  h,
  Host,
  Prop,
  Element,
  State,
  Method,
} from '@stencil/core';
import classNames from 'classnames';
import statusNote from '../../utils/status-note';

@Component({
  tag: 'scale-alertbox',
  styleUrl: 'alertbox.css',
  shadow: true,
})
export class Alertbox {
  // Color to variant rename
  @Prop() variant?: 'informational' | 'success' | 'warning' | 'error' =
    'informational';
  @Prop({ reflect: true }) hasClose?: boolean = false;
  @Prop({ reflect: true }) opened: boolean;
  @Prop() timeout?: boolean | number = false;
  /** (optional) aria-label attribute */
  @Prop() ariaLabel: string;
  /** (optional) aria-description attribute */
  @Prop() ariaDescription: string;
  @State() content: boolean = true;
  @Element() hostElement: HTMLElement;

  defaultTimeout = 3000;
  title = 'Missing Title';

  componentDidLoad() {
    this.content = !!this.hostElement.querySelector("p[slot='text']");
    this.handleSlotAccessibility();
  }

  handleSlotAccessibility() {
    let headerText = '';
    let mainText = '';

    try {
      headerText = this.hostElement.querySelector("p[slot='header']").innerHTML;
    } catch (err) {}
    try {
      mainText = this.hostElement.querySelector("p[slot='text']").innerHTML;
    } catch (err) {}

    if (headerText !== '') {
      this.hostElement.shadowRoot
        .querySelector('.alertbox__container-header')
        .setAttribute('aria-label', headerText);
    } else {
      this.hostElement.shadowRoot
        .querySelector('.alertbox__container-header')
        .setAttribute('aria-label', this.title);
    }

    if (mainText !== '') {
      this.hostElement.shadowRoot
        .querySelector('.alertbox__content')
        .setAttribute('aria-description', mainText);
    }
  }

  connectedCallback() {
    statusNote({ source: this.hostElement, type: 'warn' });
  }

  /** Alert method: open() */
  @Method()
  async open() {
    this.opened = true;
  }

  handleIcons() {
    if (this.variant) {
      switch (this.variant) {
        case 'success':
          return (
            <scale-alertbox-svg
              class="alertbox__icon-success"
              accessibility-title="success"
            />
          );
        case 'informational':
          return (
            <scale-icon-alert-information
              class="alertbox__icon-information"
              accessibility-title="information"
            />
          );
        case 'error':
        case 'warning':
          return (
            <scale-icon-alert-error
              class="alertbox__icon-error"
              accessibility-title="error"
            />
          );
      }
    }
    return;
  }

  close = () => {
    this.opened = false;
  };

  onCloseAlertWithTimeout = () => {
    if (this.timeout !== false) {
      if (typeof this.timeout === 'string' && !isNaN(this.timeout)) {
        setTimeout(this.close, this.timeout);
      } else {
        setTimeout(this.close, this.defaultTimeout);
      }
    } else {
      return null;
    }
  };

  render() {
    this.onCloseAlertWithTimeout();

    if (!this.opened) {
      return null;
    }

    return (
      <Host>
        <div
          part={this.getBasePartMap()}
          class={this.getCssClassMap()}
          aria-label={this.ariaLabel}
          tabindex="0"
        >
          <div part="container" class="alertbox__container">
            {this.handleIcons()}
            <header part="header" class="alertbox__container-header">
              <p>
                <slot name="header">{this.title}</slot>
              </p>

              {this.hasClose && (
                <scale-icon-action-circle-close
                  tabindex="0"
                  class="alertbox__icon-close"
                  onClick={() => {
                    this.close();
                  }}
                  onKeyDown={(e) => {
                    if (e.key === 'Enter') {
                      this.close();
                    }
                  }}
                  accessibility-title="close"
                />
              )}
            </header>
          </div>
          {this.content && (
            <p part="content" class="alertbox__content">
              <slot name="text" />
            </p>
          )}
        </div>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const name = 'alertbox';
    const prefix = mode === 'basePart' ? '' : `${name}--`;

    return classNames(name, this.variant && `${prefix}variant-${this.variant}`);
  }
}
