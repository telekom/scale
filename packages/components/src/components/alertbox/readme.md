# scale-alertbox



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description                           | Type                                                   | Default           |
| ----------------- | ------------------ | ------------------------------------- | ------------------------------------------------------ | ----------------- |
| `ariaDescription` | `aria-description` | (optional) aria-description attribute | `string`                                               | `undefined`       |
| `ariaLabel`       | `aria-label`       | (optional) aria-label attribute       | `string`                                               | `undefined`       |
| `hasClose`        | `has-close`        |                                       | `boolean`                                              | `false`           |
| `opened`          | `opened`           |                                       | `boolean`                                              | `undefined`       |
| `timeout`         | `timeout`          |                                       | `boolean \| number`                                    | `false`           |
| `variant`         | `variant`          |                                       | `"error" \| "informational" \| "success" \| "warning"` | `'informational'` |


## Methods

### `open() => Promise<void>`

Alert method: open()

#### Returns

Type: `Promise<void>`




## Shadow Parts

| Part          | Description |
| ------------- | ----------- |
| `"container"` |             |
| `"content"`   |             |
| `"header"`    |             |


## Dependencies

### Depends on

- [scale-icon-action-success](../icons/action-success)
- [scale-icon-alert-information](../icons/alert-information)
- [scale-icon-alert-error](../icons/alert-error)
- [scale-icon-action-circle-close](../icons/action-circle-close)

### Graph
```mermaid
graph TD;
  scale-alertbox --> scale-icon-action-success
  scale-alertbox --> scale-icon-alert-information
  scale-alertbox --> scale-icon-alert-error
  scale-alertbox --> scale-icon-action-circle-close
  style scale-alertbox fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
