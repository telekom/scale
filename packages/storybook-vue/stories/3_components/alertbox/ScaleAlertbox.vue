<template>
  <scale-alertbox
    :variant="variant"
    :has-close="hasClose"
    :timeout="timeout"
    :opened="opened"
  >
    <slot name="header"></slot>
    <slot name="text"></slot>
  </scale-alertbox>
</template>
<script>
export default {
  props: {
    variant: { type: 'error' | 'warning'| 'success'| 'informational', default: 'informational' },
    hasClose: { type: Boolean, default: false },
    opened: { type: Boolean },
    timeout: { type: Boolean | Number, default: false },
  },
};
</script>