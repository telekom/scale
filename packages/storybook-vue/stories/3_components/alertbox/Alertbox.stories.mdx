import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs/blocks';
import ScaleAlertbox from './ScaleAlertbox.vue';

<Meta
  title="Beta Components/Alertbox"
  component={ScaleAlertbox}
  argTypes={{
    variant: {
      defaultValue: 'informational',
      options: ['error', 'warning', 'success', 'informational'],
      control: {
        type: 'select',
      },
    },
    hasClose: {
      defaultValue: false,
      options: [true, false],
      control: {
        type: 'select',
      },
    },
    opened: {
      options: [true, false],
      control: {
        type: 'select',
      },
    },
    timout: {
      defaultValue: false,
      options: [false, true, Number],
      control: {
        type: 'select',
      },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleAlertbox },
  props: {
    ...ScaleAlertbox.props,
  },
  template: `
    <scale-alertbox   
     :variant="variant"
     :hasClose="hasClose"
     :timeout="timeout"
     :opened="opened"
    >
    </scale-alertbox>
  `,
});

export const TemplateSlot = (args, { argTypes }) => ({
  components: { ScaleAlertbox },
  props: {
    ...ScaleAlertbox.props,
  },
  template: `
    <scale-alertbox   
     :variant="variant"
     :hasClose="hasClose"
     :timeout="timeout"
     :opened="opened"
    >
    <template v-slot:header>
     <p slot="header">This is my title</p>
    </template>
    <template v-slot:text>
     <p slot="text">This is my message to you</p>
    </template>
  `,
});

## Standard

### All default values

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      opened: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-alertbox opened> </scale-alertbox>
```

## Scoped CSS variables

```css
:host {
  /* Alertbox styling */
  --alertbox-border-radius: 8px;
  --alertbox-border: 1px solid var(--scl-color-white);
  --alertbox-width: 100%;
  --alertbox-font-header: var(--scl-fonts-family-sans);
  --alertbox-font-size-header: 16px;
  --alertbox-font-text: var(--scl-font-weight-regular);
  --alertbox-font-size-text: 16px;
  --variant-error: hsla(353.79999999999995, 100%, 88.6%, 0.35);
  --variant-warning: hsla(33.5, 100%, 75.1%, 0.3);
  --variant-informational: hsla(190.20000000000005, 100%, 92%, 0.75);
  --variant-success: hsl(98.80000000000001, 49.7%, 61%, 0.2);
}
```

## Variants

### Warning

<Canvas withSource="none">
  <Story
    name="Warning"
    args={{
      opened: true,
      variant: 'warning',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox variant="warning" opened> </scale-alertbox>
```

### Error

<Canvas withSource="none">
  <Story
    name="Error"
    args={{
      opened: true,
      variant: 'error',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox variant="error" opened> </scale-alertbox>
```

### Success

<Canvas withSource="none">
  <Story
    name="Success"
    args={{
      opened: true,
      variant: 'success',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox variant="success" opened> </scale-alertbox>
```

### Informational

<Canvas withSource="none">
  <Story
    name="Informational"
    args={{
      opened: true,
      variant: 'informational',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox variant="informational" opened> </scale-alertbox>
```

## Close Button

<Canvas withSource="none">
  <Story
    name="Close Button"
    args={{
      opened: true,
      hasClose: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox has-close opened> </scale-alertbox>
```

## Timeout

### Set to close in 10 seconds. Default - when true: 3 seconds

<Canvas withSource="none">
  <Story
    name="Timeout"
    args={{
      opened: true,
      timeout: 10000,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox timeout="10000" opened> </scale-alertbox>
```

## Titel & Message

<Canvas withSource="none">
  <Story
    name="Titel-Message"
    args={{
      opened: true,
      hasClose: true,
    }}
  >
    {TemplateSlot.bind({})}
  </Story>
</Canvas>

```html
<scale-alertbox opened>
  <p slot="header">This is my title</p>
  <p slot="text">This is my message to you.</p>
</scale-alertbox>
```
