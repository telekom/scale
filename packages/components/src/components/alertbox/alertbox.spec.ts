import { newSpecPage } from '@stencil/core/testing';
// import { remove } from 'lodash';
import { Alertbox } from './alertbox';

describe('Alertbox', () => {
  let element;

  beforeEach(async () => {
    element = new Alertbox();
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Alertbox],
      html: `<scale-alertbox opened="true">Label</scale-alertbox>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.variant = 'warning';
    expect(element.getCssClassMap()).toContain('variant-warning');
  });

  it('should test the close() fucntion', () => {
    expect(element.opened).toBe(undefined);
    element.close();
    expect(element.opened).toBe(false);
  });

  it('handle variant prop', async () => {
    const page = await newSpecPage({
      components: [Alertbox],
      html: `<scale-alertbox opened="true">Label</scale-alertbox>`,
    });
    page.root.variant = 'error';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.root.variant = 'warning';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.root.variant = 'informational';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.root.variant = 'success';
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should reflect attributes/props', async () => {
    const page = await newSpecPage({
      components: [Alertbox],
      html: `<scale-alertbox
                variant ="warning"
                has-close ="true"
                content ="false">
              </scale-alertbox>`,
    });

    expect(page.rootInstance.variant).toBe('warning');
    expect(page.rootInstance.hasClose).toBe(true);
    expect(page.rootInstance.content).toBe(false);
  });

  it('should emit onClick and set opened to false', async () => {
    const page = await newSpecPage({
      components: [Alertbox],
      html: `<scale-alertbox opened="true" has-close="true">Label</scale-alertbox>`,
    });
    const icon = page.root.shadowRoot.querySelector('.alertbox__icon-close');
    await page.waitForChanges();
    icon.dispatchEvent(new Event('click'));
    await page.waitForChanges();
    expect(page.rootInstance.opened).toBe(false);
  });

  it('should set timeout of the FIRST argument on function onCloseAlertWithTimeout()', async () => {
    const page = await newSpecPage({
      components: [Alertbox],
      html: `<scale-alertbox opened="true" >Label</scale-alertbox>`,
    });
    page.rootInstance.defaultTimeout = 1;
    page.rootInstance.timeout = true;
    await page.waitForChanges();
    expect(page.rootInstance.defaultTimeout).toBe(1);
    setTimeout(() => {
      expect(page.root.opened).toEqual(false);
    }, 1);
  });
});
