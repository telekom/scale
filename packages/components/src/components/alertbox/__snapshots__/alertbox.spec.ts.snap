// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alertbox handle variant prop 1`] = `
<scale-alertbox opened="">
  <mock:shadow-root>
    <div class="alertbox alertbox--variant-error" part="alertbox variant-error">
      <div class="alertbox__container" part="container">
        <scale-icon-alert-error accessibility-title="error" class="alertbox__icon-error"></scale-icon-alert-error>
        <header class="alertbox__container-header" part="header">
          <p>
            <slot name="header">
              Missing Title
            </slot>
          </p>
        </header>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-alertbox>
`;

exports[`Alertbox handle variant prop 2`] = `
<scale-alertbox opened="">
  <mock:shadow-root>
    <div class="alertbox alertbox--variant-warning" part="alertbox variant-warning">
      <div class="alertbox__container" part="container">
        <scale-icon-alert-error accessibility-title="error" class="alertbox__icon-error"></scale-icon-alert-error>
        <header class="alertbox__container-header" part="header">
          <p>
            <slot name="header">
              Missing Title
            </slot>
          </p>
        </header>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-alertbox>
`;

exports[`Alertbox handle variant prop 3`] = `
<scale-alertbox opened="">
  <mock:shadow-root>
    <div class="alertbox alertbox--variant-informational" part="alertbox variant-informational">
      <div class="alertbox__container" part="container">
        <scale-icon-alert-information accessibility-title="information" class="alertbox__icon-information"></scale-icon-alert-information>
        <header class="alertbox__container-header" part="header">
          <p>
            <slot name="header">
              Missing Title
            </slot>
          </p>
        </header>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-alertbox>
`;

exports[`Alertbox handle variant prop 4`] = `
<scale-alertbox opened="">
  <mock:shadow-root>
    <div class="alertbox alertbox--variant-success" part="alertbox variant-success">
      <div class="alertbox__container" part="container">
        <scale-icon-action-success accessibility-title="success" class="alertbox__icon-success"></scale-icon-action-success>
        <header class="alertbox__container-header" part="header">
          <p>
            <slot name="header">
              Missing Title
            </slot>
          </p>
        </header>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-alertbox>
`;

exports[`Alertbox should match snapshot 1`] = `
<scale-alertbox opened="">
  <mock:shadow-root>
    <div class="alertbox alertbox--variant-informational" part="alertbox variant-informational">
      <div class="alertbox__container" part="container">
        <scale-icon-alert-information accessibility-title="information" class="alertbox__icon-information"></scale-icon-alert-information>
        <header class="alertbox__container-header" part="header">
          <p>
            <slot name="header">
              Missing Title
            </slot>
          </p>
        </header>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-alertbox>
`;
